web/src/user_status.ts

```diff
@@ -0,0 +1,144 @@
+import {z} from "zod";
+
+import * as channel from "./channel";
+import * as emoji from "./emoji";
+import type {EmojiRenderingDetails} from "./emoji";
+import {user_settings} from "./user_settings";
+
+export type UserStatus = z.infer<typeof user_status_schema>;
+export type UserStatusEmojiInfo = EmojiRenderingDetails & {
+    emoji_alt_code?: boolean;
+};
+
+type UserStatusEvent = {
+    id: number;
+    type: "user_status";
+    user_id: number;
+    away?: boolean;
+    status_text?: string;
+} & (
+    | {
+          emoji_name: string;
+          emoji_code: string | number;
+          reaction_type: string;
+      }
+    | {
+          emoji_name: undefined;
+      }
+);
+
+const user_status_schema = z.union([
+    z.object({
+        status_text: z.string().optional(),
+        emoji_name: z.string(),
+        emoji_code: z.string().or(z.number()),
+        reaction_type: z.string(),
+        away: z.boolean().optional(),
+    }),
+    z.object({
+        emoji_name: z.undefined(),
+        status_text: z.string().optional(),
+        away: z.boolean().optional(),
+    }),
+]);
+
+const user_status_param_schema = z.record(z.string(), user_status_schema);
+
+const user_info = new Map<number, string>();
+const user_status_emoji_info = new Map<number, UserStatusEmojiInfo>();
+
+export function server_update_status(opts: {
+    status_text?: string;
+    emoji_name?: string;
+    emoji_code?: string | number;
+    reaction_type?: string;
+    success?: () => void;
+}): void {
+    void channel.post({
+        url: "/json/users/me/status",
+        data: {
+            status_text: opts.status_text,
+            emoji_name: opts.emoji_name,
+            emoji_code: opts.emoji_code,
+            reaction_type: opts.reaction_type,
+        },
+        success() {
+            if (opts.success) {
+                opts.success();
+            }
+        },
+    });
+}
+
+export function server_invisible_mode_on(): void {
+    void channel.patch({
+        url: "/json/settings",
+        data: {
+            presence_enabled: false,
+        },
+    });
+}
+
+export function server_invisible_mode_off(): void {
+    void channel.patch({
+        url: "/json/settings",
+        data: {
+            presence_enabled: true,
+        },
+    });
+}
+
+export function get_status_text(user_id: number): string | undefined {
+    return user_info.get(user_id);
+}
+
+export function set_status_text(opts: {user_id: number; status_text: string}): void {
+    if (!opts.status_text) {
+        user_info.delete(opts.user_id);
+        return;
+    }
+
+    user_info.set(opts.user_id, opts.status_text);
+}
+
+export function get_status_emoji(user_id: number): UserStatusEmojiInfo | undefined {
+    return user_status_emoji_info.get(user_id);
+}
+
+export function set_status_emoji(opts: UserStatusEvent): void {
```

Comment:
Shouldn't this function be parsing the `zod` schema, since it's processing data from the server via `server_events_dispathch`?

Reply:
Yeah ideally we should parse these events as well, fixed that.